packages-exp/auth-exp/src/core/auth/auth_impl.ts

```diff
@@ -68,25 +75,73 @@ export class AuthImpl implements Auth {
   ) {}
 
   _initializeWithPersistence(
-    persistenceHierarchy: Persistence[]
+    persistenceHierarchy: Persistence[],
+    popupRedirectResolver?: externs.PopupRedirectResolver
   ): Promise<void> {
     return this.queue(async () => {
       this.persistenceManager = await PersistenceUserManager.create(
         this,
         persistenceHierarchy
       );
 
-      const storedUser = await this.persistenceManager.getCurrentUser();
-      // TODO: Check redirect user, if not redirect user, call refresh on stored user
-      if (storedUser) {
-        await this.directlySetCurrentUser(storedUser);
-      }
+      await this.initializeCurrentUser(popupRedirectResolver);
 
       this._isInitialized = true;
       this.notifyAuthListeners();
     });
   }
 
+  private async initializeCurrentUser(
+    popupRedirectResolver?: externs.PopupRedirectResolver
+  ): Promise<void> {
+    const storedUser = await this.assertedPersistence.getCurrentUser();
+    if (!storedUser) {
+      return this.directlySetCurrentUser(storedUser);
+    }
+
+    if (!storedUser._redirectEventId) {
+      // This isn't a redirect user, we can reload and bail
+      return this.reloadAndSetCurrentUserOrClear(storedUser);
+    }
+
+    assert(popupRedirectResolver, this.name, AuthErrorCode.ARGUMENT_ERROR);
+    const resolver: PopupRedirectResolver = _getInstance(popupRedirectResolver);
+
+    this.redirectPersistenceManager = await PersistenceUserManager.create(
+      this,
+      [_getInstance(resolver._redirectPersistence)],
+      _REDIRECT_USER_KEY_NAME
+    );
+
+    this.redirectUser = await this.redirectPersistenceManager.getCurrentUser();
+
+    // If the redirect user's event ID matches the current user's event ID,
+    // DO NOT reload the current user, otherwise they'll be cleared from storage.
+    // This is important for the reauthenticateWithRedirect() flow.
+    if (
+      this.redirectUser &&
+      this.redirectUser._redirectEventId === storedUser._redirectEventId
+    ) {
+      return this.directlySetCurrentUser(storedUser);
+    }
+
+    return this.reloadAndSetCurrentUserOrClear(storedUser);
+  }
+
+  private async reloadAndSetCurrentUserOrClear(user: User): Promise<void> {
+    try {
+      await _reloadWithoutSaving(user);
+    } catch (e) {
+      if (e.code !== `auth/${AuthErrorCode.NETWORK_REQUEST_FAILED}`) {
+        // User's token has expired. Log them out and remove them from storage
+        await this.directlySetCurrentUser(null);
```

Comment:
can't you just return the promise instead of awaiting?

Reply:
Done